Job History

adamfitzgibbon opened the pull request:
This pull request is for merging the updates made for the job history functionality, as requested in #17. The latest picture commented in that issue is where the page is right now, but I will update with pictures as fixes are made. Note: I left the json file with two fake jobs so that you can run it and see something in the history, but for a new user it would probably be blank by default.

Objectives yet to be completed:

- [x] Fix visual bugs (vertical misalignment, expand button color)
- [x] Transfer the selected job object to jobSubmission and populate the form with it
- [x] If a new job is created, add it to the json file upon submission. If a job is loaded, update the timestamp upon submission

- [x] Fix job submission callbacks so the job is correctly uploaded, and then run
- [x] Implement success/failure notifications
- [x] Implement a delete button so that the user can delete the history of very old jobs.

adamfitzgibbon commented:
Alright this little bit of styling has had me scratching my head for a couple days now and Google hasn't been able to figure it out for me. My problem is that it is styled correctly before I pull the date and button to the right as seen here:
![screen shot 2016-02-09 at 4 32 05 pm](https://cloud.githubusercontent.com/assets/8363569/12932930/1737e8c6-cf4b-11e5-8424-fe7e53ddd6f2.png)

But as soon as I try to align it to the right with the pull-right class, it gets sucked into the div below it:
![screen shot 2016-02-09 at 4 31 28 pm](https://cloud.githubusercontent.com/assets/8363569/12932961/36f6ea18-cf4b-11e5-8319-58bd25c47cc7.png)

It really dislikes it when I try to use grid indents, too. 

Searching around, the clearfix class seems to often be an answer to weird float issues, but it doesn't seem to be working here. Do you know of any other way to align two elements to the right that I might try?

djw8605 commented:
When you say "grid indents", do you mean like bootstrap col-md-offset-8? https://getbootstrap.com/css/#grid-offsetting

I wonder if you could put the left and right in a grid container and just put the right side in like a col-md-offset-10 (or something like that)

adamfitzgibbon commented:
Yeah, I meant offsets. That could work. The nearest grid container is also quite a few divs aways so perhaps if I made just the panel header its own grid container it wouldn't seem like the positioning is so wacky and I could position both the left and the right elements like you said?

djw8605 commented:
Yeah, I would make each panel it's own container.  With the header as the first row, and the expandable section as the second row.  Then use the grid offsets to put it to the right.

adamfitzgibbon commented:
![screen shot 2016-02-10 at 4 58 51 pm](https://cloud.githubusercontent.com/assets/8363569/12964791/a73f338a-d017-11e5-8a34-d3685a8c0095.png)
Here is the page after touchups. I decided to move the date to the left as it visually makes more sense I think. Still think the expand button could be improved a bit; maybe a triple bar expand symbol instead of words? I'll move on to making it actually work unless you have further suggestions on the looks.

djw8605 commented:
Can you space the date a little more to the right.  There should be more space between the title and date.

adamfitzgibbon commented:
![screen shot 2016-02-11 at 3 27 03 pm](https://cloud.githubusercontent.com/assets/8363569/12991212/ee832b36-d0d3-11e5-949b-ed2218de6185.png)
How's this?

djw8605 commented:
Looks great.  Can you put down arrows on either side of the "click to expand" text?

adamfitzgibbon commented:
Yeah. Do you think having the triple bar expand symbol that bootstrap often likes to use instead would be too minimalistic?

djw8605 commented:
Eh, that implies a hamburger menu, right?  Is that what you are referring to? 

![](https://cdn4.iconfinder.com/data/icons/wirecons-free-vector-icons/32/menu-alt-512.png)

adamfitzgibbon commented:
Yes that is. They seem to like to use that for things like expanding navbars for mobile layouts, but maybe that is a little too vague?

djw8605 commented:
Yeah, I like the arrows.

-Derek

> On Feb 11, 2016, at 3:35 PM, Adam Fitzgibbon <notifications@github.com> wrote:
> 
> Yes that is. They seem to like to use that for things like expanding navbars for mobile layouts, but maybe that is a little too vague?
> 
> —
> Reply to this email directly or view it on GitHub.
>

adamfitzgibbon commented:
![screen shot 2016-02-11 at 3 45 53 pm](https://cloud.githubusercontent.com/assets/8363569/12991827/bc96b02c-d0d6-11e5-90b7-c147109cf86c.png)

Alright, I think having the double arrow looks good. I'm going to read into creating a service to pass the job object to the next page's controller.

djw8605 commented on src/html/welcome.html:
```diff
@@ -3,12 +3,12 @@
   <div class="navbar-header">
     <a class="navbar-brand">HCC Go</a>
   </div>
-  
```
You have a lot of white space changes.  Can you remove them?  Look into the --ignore-whitespace option for `git`.  Also, you can you remove the white space changes from this pull request?

adamfitzgibbon commented:
I've found a command for setting an alias for this command. Is that implying that I'd have to include it in every commit? Also, I'm having trouble finding resources on writing to a file. Any suggestions on how I'd update the local json file for history?

djw8605 commented:
You will probably have to write to the file with node.js filesystem stuff.  Read in the Json, modify the data, write out the json...

adamfitzgibbon commented:
I've pushed my most recent updates. Currently the jobhistory is "working" as long as you don't recompile the application. I'll open up a separate pull request that addresses this by creating a general app data folder for HCCGo that we can use to store any user data.

adamfitzgibbon commented:
The above commit addresses issue #27. Now the application can be rebuilt and the user's job history persists as the json file is stored somewhere on their system, based on the OS. 

TODO: Fix job submission callbacks so the job is correctly uploaded, and then run. Implement success/failure notifications. Implement a delete button so that the user can delete the history of very old jobs.

djw8605 commented:
daae64b looks good.  

Here, I'll fix the todo:

- [ ] Fix job submission callbacks so the job is correctly uploaded, and then run
- [ ] Implement success/failure notifications
- [ ] Implement a delete button so that the user can delete the history of very old jobs.

adamfitzgibbon commented:
The job submission via callbacks is now functioning and the submit success/failure notifications should also be working. No progress bar or loading animations yet. Will do that after adding a delete button to history page.

djw8605 commented:
I don't see a commit for the callbacks, still incoming?

adamfitzgibbon commented:
Yeah, I was a little slow on that, here it is.

adamfitzgibbon commented:
Do you think the user should be prompted with a confirmation dialogue when they click the delete button when they're trying to delete a job from their job history? (instead of deleting the history with one click)

djw8605 commented:
Yes

-Derek

> On Mar 14, 2016, at 5:29 PM, Adam Fitzgibbon <notifications@github.com> wrote:
> 
> Do you think the user should be prompted with a confirmation dialogue when they click the delete button when they're trying to delete a job from their job history? (instead of deleting the history with one click)
> 
> —
> Reply to this email directly or view it on GitHub.
>

adamfitzgibbon commented:
Added a delete button (with confirmation) that allows the user to permanently remove the job from history. 
![screen shot 2016-04-01 at 4 35 05 pm](https://cloud.githubusercontent.com/assets/8363569/14220776/cff94260-f827-11e5-809f-89cce0c34f0d.png)

djw8605 commented:
Looks great, few comments.

1. The click to expand to change to click to hide when it its expanded.
2. When I load a file, I get this error in the script debugger:
   ![image](https://cloud.githubusercontent.com/assets/79268/14232866/eb6d748e-f97c-11e5-9f57-2ed61c264ace.png)

adamfitzgibbon commented:
Yeah I see a lot of those errors being thrown on any of the pages of the app. I'm not really sure what causes them but they don't seem to actually be negatively affecting anything. My guess is I'm doing something wrong with the angular asynchronous stuff? How would I even go about debugging that though, it doesn't seem to point to any actual non-library file.

djw8605 commented:
Yeah, It's difficult to debug.  I'm not sure.  Do you call `$scope.$apply` anywhere?  That could be the cause of it, calling it at the wrong times.

djw8605 commented:
The logic in the jobSubmission is wrong.  It claims a successful submission after the file is transferred.  Then it goes to a different location, therefore garbage collecting the objects before the actual job is able to submit.  Looks like you need to reorder and have more brackets!

adamfitzgibbon commented:
I think I'm misunderstanding how .then() works then. I was under the impression that .then() only is called after it gets the return back from submitJob which is the return from the callback of uploadJobFile. Am I using it incorrectly? My understanding on the deferred stuff was all slightly sketchy.

djw8605 commented:
Let me comment inline.  Basically you didn't call `then`on the `submitJob` function.
```javascript
// Send data to ConnectionService for file upload
    connectionService.uploadJobFile(jobFile, job.location).then(function(data) {
      // Executed when job file is successfully uploaded.
      return connectionService.submitJob(job.location);
    }).then(function() { //success
     // Also when job file is successfully uploaded
      toastr.success('Your job was succesfully submitted to the cluster!', 'Job Submitted!', {
        closeButton: true
      });
      $location.path("cluster/" + $scope.params.clusterId);
    }, function() { // error
     // When there is an error uploading the job file.
      toastr.error('There was an error in submitting your job to the cluster!', 'Job Submission Failed!', {
        closeButton: true
      });
    });
```

adamfitzgibbon commented:
Okay I'll look into fixing that when I get the chance.

adamfitzgibbon commented:
Looking back at this today, isn't this fixed by just moving the second then inside the first so that it is attached to submitJob? 
```javascript
connectionService.uploadJobFile(jobFile, job.location).then(function(data) {
  return connectionService.submitJob(job.location).then(function() { //success
        toastr.success('Your job was succesfully submitted to the cluster!', 'Job Submitted!', {
          closeButton: true
        });
        $location.path("cluster/" + $scope.params.clusterId);
      }, function() { // error
        toastr.error('There was an error in submitting your job to the cluster!', 'Job Submission Failed!', {
          closeButton: true
        });
      });
    });
```

djw8605 commented:
Yes, but you should probably catch all possible errors.  Error on file transfer isn't caught in the above.

adamfitzgibbon commented:
I fixed the calls and added a second argument to the inner then so that it should send an error notification when there is a problem. For some reason the notification wasn't getting sent when I tried it with an invalid file path so I'll have to look more into why that second then argument isn't getting called, but for now, the callbacks should at least be in order.

djw8605 commented:
Great, let me know when you are ready to test.

adamfitzgibbon commented:
I just noticed that I'm still getting that $digest error upon navigating to the Job Submission page. Any other reasons this could happen? I've only been making changes on the actual job submission, so I don't see how it could affect anything that happens on page load.

djw8605 commented on src/js/jobSubmissionCtrl.js:
```diff
@@ -71,29 +94,42 @@ jobSubmissionModule.controller('jobSubmissionCtrl', ['$scope', '$log', '$timeout
+
+    }
     selectize.refreshOptions(false);
     selectize.refreshItems();
```
Don't refresh the options and items here.  They are refreshed above.  This will cause a digest error.

adamfitzgibbon commented:
Ah, thanks. I must have overlooked that on the merge.

adamfitzgibbon commented:
Okay I believe I've fixed the error catching issue for job submission, though I'm sure I missed some unhappy paths. But it now appropriately displays an error message when it is not successful in uploading the file. When correct filepaths are given it correctly displays the success message, redirects to the home page, and runs the job.
